Better safeguard against incompatible builtin handles between dialects

clonker opened the pull request:
Inspired from the remarks by @cameel in https://github.com/ethereum/solidity/pull/15952#discussion_r2003884305, I have refactored `EVMDialect` a bit so that there is a class for generating the builtin vector that will automatically insert blanks if conditions aren't met and not give direct access to the underlying raw datastructure.
Also, I have added a unit test which takes the default dialect (without EOF) as well as the latest dialect with EOF and tests whether the populated builtin functions with their handles contained in the respective dialects are preserved when compared to all other dialects with and without object access.

clonker committed:
First define _all_ EVM functions, then declare them based on EVM version and configuration

clonker committed:
EVMDialect: add test that enforces builtin handle compatibility

Between current default dialect as well as latest dialect and all other dialects.

cameel commented on libyul/backends/evm/EVMDialect.cpp:
```diff
@@ -195,91 +196,123 @@ std::set<std::string, std::less<>> createReservedIdentifiers(langutil::EVMVersio
 	{
-		builtins.emplace_back(createIfObjectAccess("datasize", 1, 1, SideEffects{}, ControlFlowSideEffects{}, {LiteralKind::String}, [](
+		builtins.addIfObjectAccess(_objectAccess, "datasize", 1, 1, SideEffects{}, ControlFlowSideEffects{}, {LiteralKind::String}, [](
 			FunctionCall const& _call,
```
The fact that now some conditions are handled here and some in `Builtins` still bothers me a bit. I think it would be best if we made the whole thing more declarative. I.e. have one place that declares the builtins along with their scopes (i.e. EVM/EOF version range and state of object access). Then have another place that loops over all builtins that decides whether to actually define them based on their scopes.

The definition loop would then be the single place that decides whether the opcode should be skipped and whether to leave an empty spot in `m_functions`.

This would be a more comprehensive solution for the problem, because we would no longer touch that definition loop when adding new opcodes. We would only touch declarations and scopes without the risk of accidentally breaking the logic.

cameel replied:
BTW, not sure "scope" is the best way to call it, but I could not think of anything better so far. Could be just "conditions", but that sounds very non-specific.

clonker replied:
After thinking about it for a while, I think I do like `Scope`!

cameel commented on libyul/backends/evm/EVMDialect.cpp:
```diff
@@ -195,91 +196,123 @@ std::set<std::string, std::less<>> createReservedIdentifiers(langutil::EVMVersio
+			_opcode != evmasm::Instruction::JUMPF &&
+			_opcode != evmasm::Instruction::DUPN &&
+			_opcode != evmasm::Instruction::SWAPN &&
+			_opcode != evmasm::Instruction::RETF &&
```
For builtins based on opcodes, this part of the condition is the declarative part. I'd move it to a separate function that returns a boolean. Then when adding a new opcode we'd only really modify that function.

In fact, I'd actually create more than one function. I think we should make it more semantic by having smaller functions expressing specific properties the excluded opcodes satisfy:
- All the DUP/SWAP/PUSH variants: low-level stack manipulation opcodes.
    - We should also use `SemanticInformation::isSwapInstruction()`/`SemanticInformation::isDupInstruction()` here.
- All the JUMP variants and `RETF`/`CALLF`: low-level control flow opcodes
- `DATALOADN`, `EOFCREATE`, `RETURNCONTRACT`: these are actually exposed, but not directly. We have builtins replacing them. We can define a separate function for them, but a nicer alternative would be to be able to declare for a builtin that it replaces an instruction and have the definition logic exclude that instruction automatically.

Executing these functions and `EVMVersion::hasOpcode()` would then be a fixed part of the definition logic and would not have to be adjusted. It would be the equivalent of checking the scope of a builtin.

clonker replied:
I have separated out the builtin function definitions into its own class which then can serve as reference database for all EVM dialects. They can iterate over it and declare that they have a specific function by simply referring to it. 
That should also reduce memory pressure a bit, as there's only ever one copy of a builtin function (ignoring these no-output builtins).
So in principle the dialect is then simply a bool field (or implicitly convertible to one at least) over all possible builtins.

cameel commented on libyul/backends/evm/EVMDialect.cpp:
```diff
@@ -195,91 +196,123 @@ std::set<std::string, std::less<>> createReservedIdentifiers(langutil::EVMVersio
+
+/// Make sure to only add builtins in a way that is consistent over EVM versions. If the order depends on the
+/// EVM version - which can easily happen using conditionals -, different dialects' builtin handles
+/// become inherently incompatible.
```
Thinking about it, the compatibility between dialects with and without object access should be more prominently documented in the header. I'd mention it for `strictAssemblyForEVMObjects()` and maybe also for the whole `EVMDialect` class.

cameel replied:
TBH we guarantee so much compatibility between them that I'm not sure we should even be calling these distinct dialects. Move like flavors of the same dialect. It seems like an orthogonal property that could apply to any dialect.

cameel commented on test/libyul/EVMDialectCompatibility.h:
```diff
@@ -0,0 +1,78 @@
+namespace solidity::yul::test
+{
+
+inline void requireBuiltinCompatibility(EVMDialect const& _dialect1, EVMDialect const& _dialect2, std::string_view const _builtin)
+{
+	if (auto const currentHandle = _dialect1.findBuiltin(_builtin))
+		if (auto const handle = _dialect2.findBuiltin(_builtin))
```
This will only ensure that builtins available in both dialects have matching IDs, which is necessary but not sufficient for full compatibility. We also need to ensure that the dialect with objects has all the builtins present in the one without.

cameel commented on test/libyul/EVMDialectCompatibility.cpp:
```diff
@@ -0,0 +1,82 @@
+	auto const& dialectToTestAgainst = evmDialectConfigurationToTest.dialect();
+	// no object access for current
+	{
+		auto const& currentDialect = EVMDialect::strictAssemblyForEVM({}, std::nullopt);
```
I'd keep it more explicit to make places where we hard-code current EVM version easier to spot.

```suggestion
		auto const& currentDialect = EVMDialect::strictAssemblyForEVM(EVMVersion{}, std::nullopt);
```

BTW, we should probably add that `EVMVersion::current()` helper. But just internally, without exposing it to the user.

clonker replied:
Yeah, done!

cameel commented on test/libyul/EVMDialectCompatibility.cpp:
```diff
@@ -0,0 +1,82 @@
+	auto const& dialectToTestAgainst = evmDialectConfigurationToTest.dialect();
+	langutil::EVMVersion latestEVMVersion = langutil::EVMVersion::allVersions().back();
+	std::optional<uint8_t> eofVersion = std::nullopt;
+	if (withEOF)
```
It's likely that we'll forget to update this if we ever add a EOF version. We should define `allEOFVersions()` (analogous to `EVMVersion::allVersions()`), so that we can just iterate over them.

cameel commented on test/libyul/EVMDialectCompatibility.cpp:
```diff
@@ -0,0 +1,82 @@
+}
+
+BOOST_DATA_TEST_CASE(
+	builtin_function_handle_compatibility_eof,
+	bdata::monomorphic::grid(
+		bdata::make(generateEVMDialectConfigurationsToTest(std::nullopt)) + bdata::make(generateEVMDialectConfigurationsToTest(1)),
```
This seems to actually be testing both EOF and non-EOF despite the name.

cameel replied:
And a much more relevant difference is that it's using the latest rather than the default EVM version. In not so distant future both current and latest will have EOF, but the fact that current is not always the latest will remain. If you replace `withEOF` check with `withEOF && evmVersion.supportsEOF()` it will be more future-proof. You can then unify the tests and make the source EVM version just another dimension.

cameel replied:
Also, I think we should have test case where we cover the "inline" -> "with objects" conversion for all dialects. Converting between distinct dialects is nice to have but these pairs are the most relevant, because for them we actually do perform such a conversion in practice.

cameel reviewed (commented):
Looks like a step in a good direction, but if we're refactoring these builtins, there's a way to do it in a way that will more completely eliminate the possibility of introducing these kinds of mistakes. We should simply separate they declarations with the logic that defines them. See comments below for details.

clonker committed:
Add EVMVersion::current() helper

clonker commented on libyul/backends/evm/EVMDialect.cpp:
```diff
@@ -195,112 +179,203 @@ std::set<std::string, std::less<>> createReservedIdentifiers(langutil::EVMVersio
+	{
+		for (auto const& [name, opcode]: evmasm::c_instructions)
+		{
+			if (isLowLevelStackManipulationInstruction(opcode))
+				continue;
```
I have to skip them already here because there are asserts in the side-effect gathering that prevent me from creating builtin functions for these instructions in the first place. We could of course remove these asserts. Or just skip them here, before creating anything. :)

clonker committed:
Separate out builtin function collection into its own class

This makes for a better separation between the declarations and the logic that defines builtins.

# Conflicts:
#	libyul/backends/evm/EVMDialect.cpp

clonker committed:
Add allEOFVersions to EVMVersion

clonker committed:
EVMVersion::allVersions is constexpr

cameel commented on libyul/backends/evm/EVMBuiltins.cpp:
```diff
@@ -0,0 +1,418 @@
+		ControlFlowSideEffects _controlFlowSideEffects,
+		std::vector<std::optional<LiteralKind>> _literalArguments,
+		std::function<void(FunctionCall const&, AbstractAssembly&, BuiltinContext&)> _generateCode
+	)
```
Too much indent:
```suggestion
	std::function<void(FunctionCall const&, AbstractAssembly&, BuiltinContext&)> _generateCode
)
```

Same for `createFunction()` lambda in `createVerbatimFunction()`.

cameel replied:
And if we're at whitespace, please leave an empty line before `private` and `public` in the class definitions.

cameel commented on libyul/backends/evm/EVMBuiltins.cpp:
```diff
@@ -0,0 +1,418 @@
+	m_functions.emplace_back(objectAccess | requiresEOF, eofcreateBuiltin());
+	m_functions.emplace_back(objectAccess | requiresEOF, returncontractBuiltin());
+
+	using namespace std::string_literals;
```
Let's keep these at the top of the file.

clonker replied:
It seems better to me if `using namespace`s are as local as possible. Less potential for conflicting definitions. I have rewritten it to just not use the string literal so your comment is resolved but still. This is a general pet-peeve of mine with the style we use in implementation units. I'd prefer if we just explicitly declare namespace(hierarchies) and the implementation lived in them.

cameel commented on libyul/backends/evm/EVMBuiltins.cpp:
```diff
@@ -0,0 +1,418 @@
+	for (auto const& [name, opcode]: evmasm::c_instructions)
+	{
+		if (
+			opcode == evmasm::Instruction::SWAPN ||
+			opcode == evmasm::Instruction::DUPN ||
+			evmasm::SemanticInformation::isSwapInstruction(opcode) ||
+			evmasm::SemanticInformation::isDupInstruction(opcode)
```
These functions already check for `SWAPN`/`DUPN`.
```suggestion
			SemanticInformation::isSwapInstruction(opcode) ||
			SemanticInformation::isDupInstruction(opcode)
```

cameel replied:
Same in `isLowLevelStackManipulationInstruction()`.

cameel replied:
Also, shouldn't you be skipping `PUSH` and `isLowLevelControlFlowInstruction()` here as well? We don't expose them in any dialect.

clonker replied:
> These functions already check for `SWAPN`/`DUPN`.
> Same in `isLowLevelStackManipulationInstruction()`.

I purposefully did it that way because of how `AssemblyItem` is designed. We have the opcode here (type `evmasm::Instruction`) which is implicitly converted to `evmasm::AssemblyItem`. We [added asserts](https://github.com/ethereum/solidity/blob/78107e344ce526695a5a989250d01812a248d536/libevmasm/AssemblyItem.h#L89-L90) to not create assembly items for `SWAPN`/`DUPN` with the implicit conversion constructor but to always use the static helper methods. This makes it necessary to test for them individually and beforehand. It is quite unfortunate.

> Also, shouldn't you be skipping `PUSH` and `isLowLevelControlFlowInstruction()` here as well? We don't expose them in any dialect.

Not if we want the builtins thing to be a field over all possible (except for swap/dup) functions - which was my intention at least.

cameel commented on libyul/backends/evm/EVMBuiltins.h:
```diff
@@ -0,0 +1,118 @@
+	static Scopes constexpr requiresEOF{1 << requiresEOFBit};
+	static Scopes constexpr requiresNonEOF{1 << requiresNonEOFBit};
+
+	std::vector<std::tuple<Scopes, BuiltinFunctionForEVM>> m_functions;
```
I initially missed the fact that this is not just definitions but also scopes. Maybe we should put that in the name:

```suggestion
	std::vector<std::tuple<Scopes, BuiltinFunctionForEVM>> m_scopesAndFunctions;
```

cameel commented on libyul/backends/evm/EVMBuiltins.h:
```diff
@@ -0,0 +1,118 @@
+	static BuiltinFunctionForEVM createVerbatimFunction(size_t _arguments, size_t _returnVariables);
+	static SideEffects sideEffectsOfInstruction(evmasm::Instruction _instruction);
+private:
+	static Scopes constexpr instruction{1 << instructionBit};
+	static Scopes constexpr replaced{1 << replacedInstructionBit};
+	static Scopes constexpr objectAccess{1 << objectAccessBit};
+	static Scopes constexpr requiresEOF{1 << requiresEOFBit};
+	static Scopes constexpr requiresNonEOF{1 << requiresNonEOFBit};
```
```suggestion
	static Scopes constexpr instructionScope{1 << instructionBit};
	static Scopes constexpr nonInstructionScope{1 << replacedInstructionBit};
	static Scopes constexpr objectAccessScope{1 << objectAccessBit};
	static Scopes constexpr eofScope{1 << requiresEOFBit};
	static Scopes constexpr nonEOFScope{1 << requiresNonEOFBit};
```

Alternatively, might be enough to place them inside `Scopes` without the suffix so that you can refer to them as e.g. `Scope::eof`, but that would conflict with functions you already have there.

clonker replied:
I prefer it without the `Scope` suffix tbh. The type already tells the rest of the story.

cameel commented on libyul/backends/evm/EVMBuiltins.cpp:
```diff
@@ -0,0 +1,418 @@
+		ranges::all_of(
+			m_functions,
+			[](std::tuple<Scopes, BuiltinFunctionForEVM> const& _builtin)
+			{ return std::get<1>(_builtin).name.substr(0, "verbatim_"s.size()) != "verbatim_"; }),
```
```suggestion
			[](std::tuple<Scopes, BuiltinFunctionForEVM> const& _builtin) {
				return std::get<1>(_builtin).name.substr(0, "verbatim_"s.size()) != "verbatim_";
			}
		),
```

cameel commented on libyul/backends/evm/EVMDialect.cpp:
```diff
@@ -195,269 +175,51 @@ std::set<std::string, std::less<>> createReservedIdentifiers(langutil::EVMVersio
 		}
+
+		builtinShouldBeAdded &= !scopes.requiresObjectAccess() || _objectAccess;
+		builtinShouldBeAdded &= !scopes.requiresEOF() || _eofVersion.has_value();
+		builtinShouldBeAdded &= !scopes.requiresNonEOF() || !_eofVersion.has_value();
```
Please assert that both flags are not set at the same time.

Same for `instruction` and `replaced`.

clonker replied:
Agree for the EOF ones, added an assert for that. Not sure why instruction and replaced should be mutually exclusive though. One means that it is the builtin belonging to an instruction, replaced means it was replaced by something else and shouldn't be added to a dialect (in favor of whatever it was replaced with).

clonker replied:
Ah you mean that `replaced` should only be set if `instruction` is set? I'm not sure if that isn't too restrictive. Happy to add it if you insist but I don't find it particularly necessary.

cameel commented on libyul/backends/evm/EVMBuiltins.h:
```diff
@@ -0,0 +1,118 @@
+public:
+	struct Scopes
+	{
+		/// whether the corresponding evm builtin function is an instruction builtin
+		bool instruction() const { return value.test(instructionBit); }
+		/// whether the corresponding evm builtin has been replaced by another builtin, ie, should be skipped
+		bool replaced() const { return value.test(replacedInstructionBit); }
```
I'm not sure we need these two. As long as an instruction matches the EVM version and EOF, we always add it.

And when we replace an instruction we do not use the original in any dialect. We can just replace it permanently in `EVMBuiltins` and not even expose the fact that it has been replaced to `EVMDialect`.

cameel replied:
I mean, it could make sense to also go the other way and just define builtins for all instructions and only have `EVMDialect` worry about choosing the ones it wants (possibly never choosing some of them). But you don't seem to be going for that here, since you don't create builtins for `SWAP` and `DUP` instructions. If you're skipping some of them, you can just as well skip all the ones we never use.

clonker replied:
My idea was indeed to just declare all of them. Then the dialect is essentially just a bool field over all functions. For `SWAP` and `DUP` I made an exception because we currently cannot define functions for them because we can't define side effects: https://github.com/ethereum/solidity/blob/68c4aa3c943ce1e238e8fc36296928edd6394f1f/libevmasm/SemanticInformation.cpp#L481

cameel commented on libyul/backends/evm/NoOutputAssembly.cpp:
```diff
@@ -207,9 +208,36 @@ void NoOutputAssembly::appendReturnContract(ContainerID)
+		}
+
+		return modifiedBuiltins;
+	}();
```
I'm not a fan of this pattern. The `()` at the end just stand out enough. And it's not just me: [Improving Readability of IIFE ](https://www.cppstories.com/2016/11/iife-for-complex-initialization/#improving-readability-of-iife).

This two especially would be perfectly fine as normal, static functions with well defined inputs and outputs and descriptive names.

clonker replied:
Conversely you'd probably be able to find people and evidence who like the pattern. I do, for example. It makes it very easy for me to see that a variable is assigned the result of an expression and the implementation of the expression is right below it. Matter of preference and what your brain can easily digest when reading it, as it is the case with many things :)

Here especially I think the function is quite localized and does not necessarily warrant a 'full' static function, the IIFE improves readability and also clarifies scoping for definition of `m_functions`. The second one is purposefully not static as it depends on the input dialect.

I have modified it so that (I hope) it is more to your liking but I want to note that I think it rather hurts readability (for me at least) and also doesn't change anything in terms of semantics.

cameel reviewed (commented):
Looks good overall, I see no big issues, but there a few minor things that could be improved or cleaned up.

Especially the `instruction` and `replaced` scopes seem redundant to me.

nikola-matic commented on libyul/backends/evm/EVMDialect.cpp:
```diff
@@ -56,55 +59,32 @@ std::tuple<size_t, size_t> constexpr verbatimIndexToArgsAndRets(size_t _index)
-	};
-	return f;
+	return
+		_instruction == evmasm::Instruction::SWAPN ||
```
Would it make sense to introduce this to `AssemblyItem::AssemblyItemType`? On the other hand, given the uncertain future of EOF, not sure it's worth the effort.

clonker replied:
it even exists as assemblyitemtype (`evmasm::SwapN` and `evmasm::DupN`) _and_ there is a conversion constructor from instruction to assembly item type. That being said, there's also this assert that sort of ensures that swapn and dupn are only created via their respective static helper functions: https://github.com/ethereum/solidity/blob/406bdfb3abc8cfe73d59f9a748d53f9445591323/libevmasm/AssemblyItem.h#L84-L91

bottom line: I have to check for the instruction beforehand to short-circuit the expression. it's a weird design around assembly items. :/

nikola-matic commented on libyul/backends/evm/NoOutputAssembly.cpp:
```diff
@@ -226,3 +236,18 @@ BuiltinFunctionForEVM const& NoOutputEVMDialect::builtin(BuiltinHandle const& _h
+	for (auto const& [_, builtin]: allBuiltins().functions())
+	{
+		auto noOutputFunction = builtin;
+		modifyBuiltinToNoOutput(noOutputFunction);
```
What exactly does this do, or rather why? I see that it pops non-literal arguments, and then just appends zeroes for returns.

clonker replied:
It is used in the compilability checker where we don't actually want to create correct code but just see if it compiles, in principle. That thing is needed in the current implementation of stack compressor / stack limit evader. I think the design could be better around no output dialect and it would be good if we didn't need it at all. But at least in the current implementation we do :)
